File: geekbang/algorithm-models/7.1- Bellman-Ford 求最短路代码模板/networkDelayTime.java
// Java
// LeetCode 743
class Solution {
    public int networkDelayTime(int[][] times, int n, int k) {
        int[] dist = new int[n + 1];
        for (int i = 1; i <= n; i++) dist[i] = (int)1e9;
        dist[k] = 0;
        for (int iteration = 1; iteration < n; iteration++) {
            boolean updated = false;
            for (int i = 0; i < times.length; i++) {
                int x = times[i][0];
                int y = times[i][1];
                int z = times[i][2];
                if (dist[y] > dist[x] + z) {
                    dist[y] = dist[x] + z;
                    updated = true;
                }
            }
            if (!updated) break;
        }
        int ans = 0;
        for (int i = 1; i <= n; i++)
            ans = Math.max(ans, dist[i]);
        if (ans == 1e9) ans = -1;
        return ans;
    }
}

File: geekbang/algorithm-models/7.1- Bellman-Ford 求最短路代码模板/networkDelayTime.go
package main

// LeetCode 743
func networkDelayTime(times [][]int, n int, k int) int {
	dist := make([]int, n+1)
	for i := 1; i <= n; i++ {
		dist[i] = 1e9
	}
	dist[k] = 0
	for iteration := 1; iteration < n; iteration++ {
		updated := false
		for i := 0; i < len(times); i++ {
			x := times[i][0]
			y := times[i][1]
			z := times[i][2]
			if dist[y] > dist[x]+z {
				dist[y] = dist[x] + z
				updated = true
			}
		}
		if !updated {
			break
		}
	}
	ans := 0
	for i := 1; i <= n; i++ {
		if ans < dist[i] {
			ans = dist[i]
		}
	}
	if ans == 1e9 {
		ans = -1
	}
	return ans
}
